processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java

```diff
@@ -186,7 +185,7 @@ protected AddToFactsResult addToFacts(
       } else {
         // We lost a race
         aggs = concurrentGet(prev);
-        parseExceptionMessages = doAggregate(metrics, aggs, rowContainer, row);
+        doAggregate(metrics, aggs, rowContainer, row, parseExceptionMessages);
```

Comment:
Nice finding. Clearing parseExceptionMessages now.

Reply:
why not just return a List<String> like before? That seems less risky than assuming the list is mutated correctly. In general, I'm not a big fan of returning values from a function using parameters because of cases like this.

Either way, this comment shouldn't be considered a blocker, just my 2 cents

Reply:
I'm not a big fan of that either, but it seems strange to me to return exception messages as a result of aggregation. I'm actually not sure if we can return early without the second aggregation when the first aggregation throws parseExceptions. I don't want to touch that part yet at least in this PR.

Reply:
Currently (before and after this PR), a row that contains partially corrupted/incorrect data is aggregated anyway (partially).
This may result in wrong statistics inference by the user due to inconsistencies between the number of rows each aggregator accumulated.
IMO, the decision of allowing this scenario should be of the user. 
But since currently, there is no way to roll-back aggregation, I don't see how this can be easily implemented.
Maybe adding parse as a preliminary step to aggregation, but this is an entirely different subject.
"return early without the second aggregation" would only aggravate this scenario, so I don't think it is the way to go either.

Reply:
Good point. I'm not sure if that behaviour is intended or not, but seems strange to me. I think all parse operations should be done in `InputSourceReader.read()` as a preliminary step so that we don't have to worry about ParseExceptions in IncrementalIndex. But yeah, this is a different subject.